### PR TITLE
Update and simplify cmake

### PR DESCRIPTION
**Pull request by DiamonDinoia**
Dear all,

I made some changes on cmake to remove the _static to finufft and cleanup other things in the process.

Changelog:
- [x] `-march=native` like on msvc
- [x] static libraries by default only (shared replaces static)
- [x] flags are detected by cmake and all the supported ones are enabled

Known issues:
- [x] move msvc march=native to a separate file
- [x] build GPU library does not depend on CPU anymore
- [x] GPU should not depend on FFTW
- [x] clang-cl declspec issue

### CODE REVIEW

**Comment by DiamonDinoia** [comment]
@lu1and10 do you know why windows clang-cl is not building?

**Comment by lu1and10** [comment]
> @lu1and10 do you know why windows clang-cl is not building?

I'm not sure, it seems too much going on with the cmake changes, need to take a look in detail.

**Comment by DiamonDinoia** [comment]
I think I fixed everything. I think we should use the matrix I use for testing python to test cmake-ci too. This way we can test all os and all compilers.

**Comment by janden** [comment]
To reduce issues with CMakeLists style, etc., would it make sense to use a tool like `cmake-format` (https://github.com/cheshirekow/cmake_format) in our pre-commit hook?

**Comment by DiamonDinoia** [comment]
@Wentzell could you have a look at this? We should incorporate namespaces and have proper packaging across all OSes, any suggestion would be precious as I have not done anything similar in the past.

**Comment by DiamonDinoia** [comment]
I run some tests to fix windows .dll build results below:

msvc shared liking:

`finufft.dll` depends on `VCOMP140D.DLL` & `KERNEL32.DLL`

I believe this is acceptable. Python ships the first so there is no issue and the second we can add it to the install notes as a requirement as this seems standard among windows users.



Issue: Finufft build shared libs: dll not found error 0xc0000135
Solution:
- [ ] Based on the way [windows loads .dll](https://stackoverflow.com/questions/23323741/building-of-executable-and-shared-library-with-cmake-runtimelinker-does-not-fin).  I think we need to copy the .dll in the tests directory
- [ ] Update install notes so that users can either update their path to contain root folder of finufft.dll (it depends on where the user installs it). Or they manually need to copy the .dll in their executable path. For cmake users they can use a snippet to have cmake copy the .dll for them that I can provide.

File: src/cuda/CMakeLists.txt [hunk 1]
```diff
@@ -1,60 +1,89 @@
...
+
+# flush denormals to zero and enable verbose PTXAS output
+set(FINUFFT_CUDA_FLAGS
+    -ftz=true
+    -fmad=true
+    -restrict
+    --extra-device-vectorization
+    $<$<CONFIG:Debug>:-G
+    -maxrregcount
+    32>)
```
[the remark this answers is not in the document]
Comment (lu1and10) [reply src/cuda/CMakeLists.txt:42]: > Does this `-maxrregcount 32` give the best performance to limit cuda kernel to 32 registers?

only for debug mode.

**Review by lu1and10 — APPROVED** [review]

**Comment by lu1and10** [comment]
One note for future: @DiamonDinoia finds that on windows with msvc cl, ducc0 **has to** compile with **/fp:fast**, otherwise some tests(`run_finufft3d_test_float`, `run_finufft3dmany_test_float`) may fail because of the resulting error is larger than the tolerance.

On the other hand, finufft on windows with msvc cl **should not** compile with flag **/fp:fast**, with /fp:fast the test `run_dumbinputs_double` will result in segfault, because /fp:fast makes [ values (NaN, +infinity, -infinity, -0.0) may not be propagated or behave strictly according to the IEEE-754 standard](https://learn.microsoft.com/en-us/cpp/build/reference/fp-specify-floating-point-behavior?view=msvc-170#fast) 

We may need to add this note in the installation doc.

**Pull request merged**